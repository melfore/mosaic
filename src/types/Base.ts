export enum Color {
  default = "default",
  inherit = "inherit",
  primary = "primary",
  secondary = "secondary",
}

export interface BaseType {
  color?: Color;
  dataCy?: string;
}

export interface BaseIntlType extends BaseType {
  labelId: string;
}

export interface LoadableType extends BaseType {
  loading?: boolean;
}
