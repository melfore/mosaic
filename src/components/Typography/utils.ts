type VariantDefaultsType = {
  [key: string]: {
    bottomSpacing: boolean;
    truncated: boolean;
  };
};

const VARIANT_DEFAULTS: VariantDefaultsType = {
  body1: {
    bottomSpacing: false,
    truncated: false,
  },
  caption: {
    bottomSpacing: false,
    truncated: false,
  },
  h5: {
    bottomSpacing: true,
    truncated: true,
  },
  h6: {
    bottomSpacing: true,
    truncated: true,
  },
  overline: {
    bottomSpacing: false,
    truncated: false,
  },
  subtitle1: {
    bottomSpacing: false,
    truncated: false,
  },
};

export const getBottomSpacing = (variant: string, bottomSpacing?: boolean): boolean =>
  bottomSpacing === undefined ? VARIANT_DEFAULTS[variant].bottomSpacing : bottomSpacing;

export const getTruncate = (variant: string, truncated?: boolean): boolean =>
  truncated === undefined ? VARIANT_DEFAULTS[variant].truncated : truncated;

export const VARIANT_COMPONENT_MAP = {
  body1: "p",
  caption: "span",
  h5: "h1",
  h6: "h2",
  overline: "h4",
  subtitle1: "h3",
};
