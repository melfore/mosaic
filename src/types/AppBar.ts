import { BaseType } from "./Base";
import { Icons } from "./Icon";

interface UserMenuItem {
  label: string;
  onClick: () => void;
}

interface ActionItem {
  icon: Icons;
  onClick: () => void;
}

export interface AppBarType extends BaseType {
  actions?: ActionItem[];
  onNavigationMenuClick?: () => void;
  onTitleClick?: () => void;
  title?: string;
  userMenu?: UserMenuItem[];
}
