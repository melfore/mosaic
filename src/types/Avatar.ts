import { LoadableType } from "./Base";
import { Icons } from "./Icon";

export enum AvatarVariant {
  default = "circle",
  rounded = "rounded",
  squared = "square",
}

export interface AvatarType extends LoadableType {
  alt?: string;
  icon?: Icons;
  src?: string;
  text?: string;
  variant?: AvatarVariant;
}
