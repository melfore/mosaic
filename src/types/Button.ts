import { MouseEvent } from "react";
import { BaseType } from "./Base";
import { Icons } from "./Icon";

export enum ButtonIconPosition {
  left = "left",
  right = "right",
}

export interface ButtonIconType {
  name: Icons;
  position?: ButtonIconPosition;
}

export enum ButtonVariants {
  contained = "contained",
  outlined = "outlined",
}

export interface ButtonType extends BaseType {
  elevated?: boolean;
  icon?: ButtonIconType;
  label?: string;
  onClick: (event: MouseEvent) => void;
  variant?: ButtonVariants;
}
