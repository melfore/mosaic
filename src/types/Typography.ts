import { BaseType } from "./Base";

export enum TypographyVariants {
  body = "body1",
  caption = "caption",
  label = "subtitle1",
  overline = "overline",
  pagetitle = "h5",
  title = "h6",
}

export interface TypographyType extends BaseType {
  bottomSpacing?: boolean;
  label?: string;
  truncated?: boolean;
  variant?: TypographyVariants;
}
