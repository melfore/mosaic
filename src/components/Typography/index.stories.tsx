import React from "react";
import {} from "@storybook/addon-actions";
import { text, select, boolean } from "@storybook/addon-knobs";
import { TypographyVariants } from "../../types/Typography";
import IntlProviderMock, { LocaleMock, MessageMock } from "../../utils/mocks/IntlProviderMock";
import { DOCS_PAGE_STRUCTURE, StoriesWrapper } from "../../utils/stories";
import Typography, { TypographyIntl } from ".";

export default {
  title: "Typography",
  component: Typography,
  parameters: {
    ...DOCS_PAGE_STRUCTURE,
  },
};

export const Canvas = () => (
  <Typography
    bottomSpacing={boolean("bottomSpacing", false)}
    label={text("label", "Typography example")}
    truncated={boolean("truncated", false)}
    variant={select("variant", TypographyVariants, TypographyVariants.body)}
  />
);

const allVariantsKeys = Object.keys(TypographyVariants);
const allVariantsValues = Object.values(TypographyVariants);
const allVariants: { [key: string]: string } = {};
allVariantsKeys.forEach((key, index) => (allVariants[key] = allVariantsValues[index]));
const getVariant = (variantKey: string) => allVariantsValues.find((vv) => vv === allVariants[variantKey]);
const allVariantsJsx = allVariantsKeys.map((variantKey) => (
  <div className="typography-wrapper">
    <Typography label={`${variantKey}`} variant={TypographyVariants.overline} />
    <Typography
      dataCy={`typography-${variantKey}`}
      label={`This is an example text`}
      variant={getVariant(variantKey)}
    />
  </div>
));

export const AllVariants = () => <StoriesWrapper>{allVariantsJsx}</StoriesWrapper>;

export const Truncated = () => (
  <StoriesWrapper>
    <div className="typography-wrapper">
      <div style={{ width: "50%" }}>
        <Typography label="Trucates with respect to parent width (resize window to make it happen)" truncated />
      </div>
      <div style={{ width: "200px" }}>
        <Typography label="Truncated all chars after 200px" truncated />
      </div>
    </div>
  </StoriesWrapper>
);

export const WithBottomSpacing = () => (
  <StoriesWrapper>
    <div className="typography-wrapper">
      <Typography label="Some text without bottom spacing" />
      <Typography bottomSpacing label="Some text with bottom spacing" />
      <Typography label="Other text" />
    </div>
  </StoriesWrapper>
);

export const WithIntl = () => (
  // IntlProviderMock simulates external IntlProvider context
  // The `TypographyIntl` exported version of `Typography` uses `labelId` prop to get a localized label.
  // When using `TypographyIntl` the prop `dataCy` can be omitted as it defaults to `labelId` value.
  <StoriesWrapper>
    <IntlProviderMock locale={select("locale", LocaleMock, LocaleMock.en)}>
      <TypographyIntl labelId={MessageMock.typography} />
    </IntlProviderMock>
  </StoriesWrapper>
);
