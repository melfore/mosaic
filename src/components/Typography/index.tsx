import React, { FC } from "react";
import MUITypography from "@material-ui/core/Typography";
import { BaseIntlType } from "../../types/Base";
import { TypographyType, TypographyVariants } from "../../types/Typography";
import withIntl from "../../utils/hocs/withIntl";
import { getBottomSpacing, getTruncate, VARIANT_COMPONENT_MAP } from "./utils";

/**
 * Typography component made on top of `@material-ui/core/Typography`.
 */
const Typography: FC<TypographyType> = ({
  bottomSpacing = undefined,
  dataCy = "typography",
  label,
  truncated = undefined,
  variant = TypographyVariants.body,
}) => {
  return (
    <MUITypography
      className={`data-cy-${dataCy}`}
      gutterBottom={getBottomSpacing(variant, bottomSpacing)}
      noWrap={getTruncate(variant, truncated)}
      variant={variant}
      variantMapping={VARIANT_COMPONENT_MAP}
    >
      {label}
    </MUITypography>
  );
};

export const TypographyIntl: FC<TypographyType & BaseIntlType> = withIntl(Typography);

export default Typography;
