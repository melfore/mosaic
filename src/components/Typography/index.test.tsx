import React from "react";
import { mount } from "enzyme";
import Typography from ".";
import { TypographyVariants } from "../../types/Typography";

const defaultProps = {};

const componentWrapper = (props = {}) => <Typography {...defaultProps} {...props} />;

describe("Typography test suite:", () => {
  it("default", () => {
    const component = componentWrapper();
    const wrapper = mount(component);
    const styledTypography = wrapper.find("Typography").childAt(0);
    expect(styledTypography.prop("gutterBottom")).toBeFalsy();
    expect(styledTypography.prop("noWrap")).toBeFalsy();
    const paragraph = styledTypography.find("p");
    expect(paragraph.prop("className")).toContain("MuiTypography-body1");
    expect(paragraph.prop("className")).toContain("data-cy-typography");
  });

  it("pagetitle", () => {
    const component = componentWrapper({ variant: TypographyVariants.pagetitle });
    const wrapper = mount(component);
    const styledTypography = wrapper.find("Typography").childAt(0);
    expect(styledTypography.prop("gutterBottom")).toBeTruthy();
    expect(styledTypography.prop("noWrap")).toBeTruthy();
    const pageTitle = styledTypography.find("h1");
    expect(pageTitle.prop("className")).toContain("MuiTypography-h5");
  });

  it("with bottom spacing", () => {
    const component = componentWrapper({ bottomSpacing: true });
    const wrapper = mount(component);
    const styledTypography = wrapper.find("Typography").childAt(0);
    expect(styledTypography.prop("gutterBottom")).toBeTruthy();
  });

  it("with custom data-cy", () => {
    const component = componentWrapper({ dataCy: "custom" });
    const wrapper = mount(component);
    const paragraph = wrapper.find("p");
    expect(paragraph.prop("className")).toContain("data-cy-custom");
  });

  it("with truncated", () => {
    const component = componentWrapper({ truncated: true });
    const wrapper = mount(component);
    const styledTypography = wrapper.find("Typography").childAt(0);
    expect(styledTypography.prop("noWrap")).toBeTruthy();
  });
});
