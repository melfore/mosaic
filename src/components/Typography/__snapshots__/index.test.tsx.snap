// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography test suite: bottomSpacing 1`] = `
<p
  className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-397282-MuiTypography-root"
  data-cy="typography"
>
  Text
</p>
`;

exports[`Typography test suite: dataCy 1`] = `
<p
  className="MuiTypography-root MuiTypography-body1 css-5wwosz-MuiTypography-root"
  data-cy="custom"
>
  Text
</p>
`;

exports[`Typography test suite: default 1`] = `
<p
  className="MuiTypography-root MuiTypography-body1 css-5wwosz-MuiTypography-root"
  data-cy="typography"
>
  Text
</p>
`;

exports[`Typography test suite: display 1`] = `
<p
  className="MuiTypography-root MuiTypography-body1 css-1sx1eu0-MuiTypography-root"
  data-cy="typography"
>
  Text
</p>
`;

exports[`Typography test suite: loading 1`] = `
<p
  className="MuiTypography-root MuiTypography-body1 css-5wwosz-MuiTypography-root"
  data-cy="typography-loading"
>
  <span
    className="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1e1x4vv-MuiSkeleton-root"
    style={
      {
        "height": undefined,
        "width": undefined,
      }
    }
  />
</p>
`;

exports[`Typography test suite: localized 1`] = `
<p
  className="MuiTypography-root MuiTypography-body1 css-5wwosz-MuiTypography-root"
  data-cy="typography"
>
  Text
</p>
`;

exports[`Typography test suite: truncated 1`] = `
<p
  className="MuiTypography-root MuiTypography-body1 MuiTypography-noWrap css-1m2dqj2-MuiTypography-root"
  data-cy="typography"
>
  Text
</p>
`;

exports[`Typography test suite: variant 1`] = `
<h2
  className="MuiTypography-root MuiTypography-h6 css-tho60i-MuiTypography-root"
  data-cy="typography"
>
  Text
</h2>
`;
